File: src/cpp/triangular.cpp
#include "ordinal.h"
#include <algorithm>
#include <set>
#include <stdexcept>
#include <vector>

// ============================================================
// Ancestor-based BMS ↔ Triangular BMS conversion
// Ported from bms_ancestor_converter_v6.py
// Supports arbitrary n-row matrices.
// ============================================================

using Column = std::vector<int>;
using Columns = std::vector<Column>;

// ── Column comparison (lexicographic) ──

static bool colEq(const Column &a, const Column &b) {
  if (a.size() != b.size())
    return false;
  for (size_t i = 0; i < a.size(); i++)
    if (a[i] != b[i])
      return false;
  return true;
}
static bool colLess(const Column &a, const Column &b) {
  size_t n = std::min(a.size(), b.size());
  for (size_t i = 0; i < n; i++)
    if (a[i] != b[i])
      return a[i] < b[i];
  return a.size() < b.size();
}
// a >= b  ⇔  !(a < b)
static bool colGeq(const Column &a, const Column &b) { return !colLess(a, b); }

/// Lexicographic compare of two column sequences (Python tuple-of-tuples style).
/// Returns <0 (a<b), 0 (a==b), >0 (a>b).
static int seqCmp(const Columns &a, const Columns &b) {
  size_t n = std::min(a.size(), b.size());
  for (size_t i = 0; i < n; i++) {
    if (colEq(a[i], b[i]))
      continue;
    return colLess(a[i], b[i]) ? -1 : 1;
  }
  if (a.size() == b.size())
    return 0;
  return a.size() < b.size() ? -1 : 1;
}

// ── Helpers ──

/// Largest 1-based row where col[row-1] > 0; 0 if all zero.
static int lastPositiveRow(const Column &col) {
  for (int i = (int)col.size() - 1; i >= 0; i--)
    if (col[i] > 0)
      return i + 1;
  return 0;
}

static Column incrementPrefix(const Column &col, int count) {
  Column r = col;
  for (int i = 0; i < count && i < (int)r.size(); i++)
    r[i]++;
  return r;
}

static Column decrementPrefix(const Column &col, int count) {
  Column r = col;
  for (int i = 0; i < count && i < (int)r.size(); i++) {
    if (r[i] == 0)
      return Column(); // sentinel failure
    r[i]--;
  }
  return r;
}

static Column incrementRow(const Column &col, int row /*1-based*/) {
  Column r = col;
  if (row >= 1 && row <= (int)r.size())
    r[row - 1]++;
  return r;
}

static Column zeroFromRow(const Column &col, int row /*1-based*/) {
  Column r = col;
  for (int i = row - 1; i < (int)r.size(); i++)
    r[i] = 0;
  return r;
}

static Column firstRowColumn(int value, int n) {
  Column r(n, 0);
  r[0] = value;
  return r;
}

/// Check if all entries in first `count` positions are > 0 (can be decremented).
static bool canDecrementPrefix(const Column &col, int count) {
  for (int i = 0; i < count && i < (int)col.size(); i++)
    if (col[i] == 0)
      return false;
  return true;
}

// ── AncestorIndex ──

class AncestorIndex {
public:
  int n;
  int columnCount;
  Columns columns;
  std::vector<std::vector<int>> parents;             // parents[row][col]; -1 = none
  std::vector<std::vector<std::set<int>>> ancestors; // ancestors[row][col]

  AncestorIndex(const Columns &cols) : columns(cols) {
    if (cols.empty()) {
      n = 0;
      columnCount = 0;
      return;
    }
    columnCount = (int)cols.size();
    n = (int)cols[0].size();

    parents.assign(n + 1, std::vector<int>(columnCount, -1));
    ancestors.assign(n + 1, std::vector<std::set<int>>(columnCount));

    // Row 0 (virtual): ancestors are all columns to the left
    for (int c = 1; c < columnCount; c++) {
      parents[0][c] = c - 1;
      for (int a = 0; a < c; a++)
        ancestors[0][c].insert(a);
    }

    // Rows 1..n
    for (int row = 1; row <= n; row++) {
      int vi = row - 1;
      for (int c = 0; c < columnCount; c++) {
        int parent = -1;
        auto &up = ancestors[row - 1][c];
        for (auto it = up.rbegin(); it != up.rend(); ++it) {
          if (cols[*it][vi] < cols[c][vi]) {
            parent = *it;
            break;
          }
        }
        parents[row][c] = parent;
        if (parent >= 0) {
          ancestors[row][c] = ancestors[row][parent];
          ancestors[row][c].insert(parent);
        }
      }
    }
  }

  bool hasAncestorColumn(int elementCol, int row, int ancestorCol) const {
    if (row < 0 || row > n || elementCol < 0 || elementCol >= columnCount || ancestorCol < 0 || ancestorCol >= columnCount)
      return false;
    return ancestors[row][elementCol].count(ancestorCol) > 0;
  }

  bool parentIsColumn(int elementCol, int row, int parentCol) const {
    if (row < 0 || row > n || elementCol < 0 || elementCol >= columnCount)
      return false;
    return parents[row][elementCol] == parentCol;
  }
};

// ============================================================
// Triangular BMS → Standard BMS
// ============================================================

Matrix triangularToBMS(const Matrix &M) {
  Columns cols;
  for (auto &c : M)
    cols.push_back(c);
  if (cols.empty())
    return {};
  int n = (int)cols[0].size();
  if (n < 2)
    return {};

  int idx = (int)cols.size() - 1;

  while (idx >= 0) {
    Column &x = cols[idx];
    int rowNminus2 = (n >= 2) ? x[n - 2] : 0;
    if (rowNminus2 > 0) {
      idx--;
      continue;
    }

    int k = lastPositiveRow(x);
    if (k + 2 > n) {
      idx--;
      continue;
    }

    Column y = incrementPrefix(x, k + 1);
    Column z = incrementPrefix(y, k + 2);

    int yIdx = idx + 1;
    int machineStart = idx + 2;

    if (yIdx >= (int)cols.size() || !colEq(cols[yIdx], y) || machineStart >= (int)cols.size() ||
        colLess(cols[machineStart], z)) {
      idx--;
      continue;
    }

    AncestorIndex ancestor(cols);
    Columns xPrime;
    int cursor = machineStart;
    int xEnd = cursor;
    int lastL = -1;
    bool lastStoppedByXParent = false;

    while (true) {
      if (cursor >= (int)cols.size() || colLess(cols[cursor], z)) {
        xEnd = cursor;
        break;
      }

      Column &t = cols[cursor];

      // Find matching rows l (0..k+1) where t[l] has ancestor in y
      int l = -1;
      for (int row = 0; row <= k + 1; row++) {
        if (ancestor.hasAncestorColumn(cursor, row, yIdx))
          l = row; // take max
      }
      if (l < 0)
        return {}; // non-standard

      bool stoppedByXParent = (l <= k) && ancestor.parentIsColumn(cursor, l + 1, idx);

      Column tPrime = decrementPrefix(t, l);
      if (tPrime.empty())
        return {};

      if (stoppedByXParent)
        tPrime = zeroFromRow(tPrime, l + 2);

      xPrime.push_back(tPrime);
      cursor++;

      lastL = l;
      lastStoppedByXParent = stoppedByXParent;

      if (stoppedByXParent) {
        xEnd = cursor;
        break;
      }
    }

    // Determine if we keep y and the original X
    bool keepCase1 = false;
    if (xEnd < (int)cols.size()) {
      Column frc = firstRowColumn(z[0], n);
      keepCase1 = colGeq(cols[xEnd], frc);
    }

    bool keepCase2 = false;
    if (lastL >= 0 && xEnd > 0 && xEnd - 1 < (int)cols.size()) {
      if (cols[xEnd - 1][lastL] == 0 && ancestor.parentIsColumn(xEnd - 1, lastL, yIdx))
        keepCase2 = true;
    }

    bool keepCase3 =
        lastStoppedByXParent && lastL + 1 < n && xEnd > 0 && (xEnd - 1) < (int)cols.size() && cols[xEnd - 1][lastL + 1] > 0;

    bool keepOriginalYX = keepCase1 || keepCase2 || keepCase3;

    if (keepOriginalYX) {
      // Keep y and x_prime after x, original X remains
      cols.insert(cols.begin() + idx + 1, xPrime.begin(), xPrime.end());
    } else {
      // Replace y..X with x_prime only
      cols.erase(cols.begin() + idx + 1, cols.begin() + xEnd);
      cols.insert(cols.begin() + idx + 1, xPrime.begin(), xPrime.end());
    }

    idx--;
  }

  return cols;
}

// ============================================================
// Standard BMS → Triangular BMS
// ============================================================

Matrix bmsToTriangular(const Matrix &M) {
  Columns cols;
  for (auto &c : M)
    cols.push_back(c);
  if (cols.empty())
    return {};
  int n = (int)cols[0].size();
  if (n < 2)
    return {};

  int idx = 0;
  int steps = 0;
  const int stepLimit = 100000;

  while (idx < (int)cols.size()) {
    steps++;
    if (steps > stepLimit)
      return {};

    Column &x = cols[idx];
    int k = lastPositiveRow(x);
    if (k >= n - 1) {
      idx++;
      continue;
    }

    Column y = incrementPrefix(x, k + 1);
    Column z = incrementRow(y, k + 2);

    int xStart = idx + 1;
    if (xStart >= (int)cols.size() || colLess(cols[xStart], z)) {
      idx++;
      continue;
    }

    int xEnd = xStart;
    while (xEnd < (int)cols.size() && colGeq(cols[xEnd], z))
      xEnd++;

    AncestorIndex ancestor(cols);
    Columns xPrime;

    for (int cursor = xStart; cursor < xEnd; cursor++) {
      Column &t = cols[cursor];

      int l = -1;
      for (int row = 0; row <= k + 1; row++) {
        if (ancestor.hasAncestorColumn(cursor, row, idx))
          l = row;
      }
      if (l < 0)
        return {};

      bool isLast = (cursor == xEnd - 1);
      if (isLast) {
        if (l < 0 || l >= n)
          return {};
        if (ancestor.parentIsColumn(cursor, l, idx) && t[l] == 0)
          l--;
      }
      if (l < 0)
        return {};

      Column tPrime = incrementPrefix(t, l);
      xPrime.push_back(tPrime);
    }

    // comparison matrix: (y, ...x_prime, (y[0]+1,0,...,0))
    // remainder: columns starting from xEnd
    Columns insertion;
    insertion.push_back(y);
    insertion.insert(insertion.end(), xPrime.begin(), xPrime.end());

    Columns comparison = insertion;
    comparison.push_back(firstRowColumn(y[0] + 1, n));

    Columns remainder(cols.begin() + xEnd, cols.end());

    // Always delete X
    cols.erase(cols.begin() + xStart, cols.begin() + xEnd);

    // Insert if comparison > remainder (lexicographic tuple-of-tuples)
    if (seqCmp(comparison, remainder) > 0)
      cols.insert(cols.begin() + xStart, insertion.begin(), insertion.end());

    idx++;
  }

  return cols;
}

File: src/rust/core/src/triangular.rs
#![allow(unused_assignments)]
//! Ancestor-based BMS ↔ Triangular BMS conversion.

use crate::Matrix;
use std::collections::BTreeSet;

type Column = Vec<i32>;
type Columns = Vec<Column>;

fn col_eq(a: &Column, b: &Column) -> bool {
    a == b
}

fn col_less(a: &Column, b: &Column) -> bool {
    let n = a.len().min(b.len());
    for i in 0..n {
        if a[i] != b[i] {
            return a[i] < b[i];
        }
    }
    a.len() < b.len()
}

fn col_geq(a: &Column, b: &Column) -> bool {
    !col_less(a, b)
}

/// Lexicographic compare of two column sequences.
fn seq_cmp(a: &Columns, b: &Columns) -> i32 {
    let n = a.len().min(b.len());
    for i in 0..n {
        if col_eq(&a[i], &b[i]) {
            continue;
        }
        return if col_less(&a[i], &b[i]) { -1 } else { 1 };
    }
    if a.len() == b.len() {
        return 0;
    }
    if a.len() < b.len() {
        -1
    } else {
        1
    }
}

/// Largest 1-based row where col[row-1] > 0; 0 if all zero.
fn last_positive_row(col: &Column) -> i32 {
    for i in (0..col.len()).rev() {
        if col[i] > 0 {
            return i as i32 + 1;
        }
    }
    0
}

fn increment_prefix(col: &Column, count: i32) -> Column {
    let mut r = col.clone();
    for i in 0..(count as usize).min(r.len()) {
        r[i] += 1;
    }
    r
}

fn decrement_prefix(col: &Column, count: i32) -> Column {
    let mut r = col.clone();
    for i in 0..(count as usize).min(r.len()) {
        if r[i] == 0 {
            return Column::new(); // sentinel failure
        }
        r[i] -= 1;
    }
    r
}

fn increment_row(col: &Column, row: i32 /*1-based*/) -> Column {
    let mut r = col.clone();
    if row >= 1 && (row as usize) <= r.len() {
        r[(row - 1) as usize] += 1;
    }
    r
}

fn zero_from_row(col: &Column, row: i32 /*1-based*/) -> Column {
    let mut r = col.clone();
    for i in ((row - 1) as usize)..r.len() {
        r[i] = 0;
    }
    r
}

fn first_row_column(value: i32, n: usize) -> Column {
    let mut r = vec![0i32; n];
    r[0] = value;
    r
}

struct AncestorIndex {
    n: usize,
    column_count: usize,
    _columns: Columns,
    parents: Vec<Vec<i32>>,            // parents[row][col]; -1 = none
    ancestors: Vec<Vec<BTreeSet<usize>>>, // ancestors[row][col]
}

impl AncestorIndex {
    fn new(cols: &Columns) -> Self {
        if cols.is_empty() {
            return AncestorIndex {
                n: 0,
                column_count: 0,
                _columns: cols.clone(),
                parents: Vec::new(),
                ancestors: Vec::new(),
            };
        }
        let column_count = cols.len();
        let n = cols[0].len();

        let mut parents = vec![vec![-1i32; column_count]; n + 1];
        let mut ancestors: Vec<Vec<BTreeSet<usize>>> = vec![vec![BTreeSet::new(); column_count]; n + 1];

        // Row 0 (virtual): ancestors are all columns to the left
        for c in 1..column_count {
            parents[0][c] = c as i32 - 1;
            for a in 0..c {
                ancestors[0][c].insert(a);
            }
        }

        // Rows 1..n
        for row in 1..=n {
            let vi = row - 1;
            for c in 0..column_count {
                let mut parent: i32 = -1;
                // iterate up (ancestors of row-1 at c) in reverse (right to left)
                let up: Vec<usize> = ancestors[row - 1][c].iter().cloned().collect();
                for &it in up.iter().rev() {
                    if cols[it][vi] < cols[c][vi] {
                        parent = it as i32;
                        break;
                    }
                }
                parents[row][c] = parent;
                if parent >= 0 {
                    let pu = parent as usize;
                    ancestors[row][c] = ancestors[row][pu].clone();
                    ancestors[row][c].insert(pu);
                }
            }
        }

        AncestorIndex {
            n,
            column_count,
            _columns: cols.clone(),
            parents,
            ancestors,
        }
    }

    fn has_ancestor_column(&self, element_col: usize, row: usize, ancestor_col: usize) -> bool {
        if row > self.n || element_col >= self.column_count || ancestor_col >= self.column_count {
            return false;
        }
        self.ancestors[row][element_col].contains(&ancestor_col)
    }

    fn parent_is_column(&self, element_col: usize, row: usize, parent_col: usize) -> bool {
        if row > self.n || element_col >= self.column_count {
            return false;
        }
        self.parents[row][element_col] == parent_col as i32
    }
}

// ============================================================
// Triangular BMS → Standard BMS
// ============================================================

pub fn triangular_to_bms(m: &Matrix) -> Matrix {
    let mut cols = m.clone();
    if cols.is_empty() {
        return Vec::new();
    }
    let n = cols[0].len();
    if n < 2 {
        return Vec::new();
    }

    let mut idx = cols.len() as i32 - 1;

    while idx >= 0 {
        let iu = idx as usize;
        let x = cols[iu].clone();
        let row_n_minus_2 = if n >= 2 { x[n - 2] } else { 0 };
        if row_n_minus_2 > 0 {
            idx -= 1;
            continue;
        }

        let k = last_positive_row(&x);
        if k + 2 > n as i32 {
            idx -= 1;
            continue;
        }

        let y = increment_prefix(&x, k + 1);
        let z = increment_prefix(&y, k + 2);

        let y_idx = iu + 1;
        let machine_start = iu + 2;

        if y_idx >= cols.len()
            || !col_eq(&cols[y_idx], &y)
            || machine_start >= cols.len()
            || col_less(&cols[machine_start], &z)
        {
            idx -= 1;
            continue;
        }

        let ancestor = AncestorIndex::new(&cols);
        let mut x_prime: Columns = Vec::new();
        let mut cursor = machine_start;
        let mut x_end = cursor;
        let mut last_l: i32 = -1;
        let mut last_stopped_by_x_parent = false;

        loop {
            if cursor >= cols.len() || col_less(&cols[cursor], &z) {
                x_end = cursor;
                break;
            }

            let t = cols[cursor].clone();

            // Find matching rows l (0..k+1) where t[l] has ancestor in y
            let mut l: i32 = -1;
            for row in 0..=(k as usize + 1) {
                if ancestor.has_ancestor_column(cursor, row, y_idx) {
                    l = row as i32; // take max
                }
            }
            if l < 0 {
                return Vec::new(); // non-standard
            }

            let stopped_by_x_parent = (l <= k) && ancestor.parent_is_column(cursor, (l + 1) as usize, iu);

            let mut t_prime = decrement_prefix(&t, l);
            if t_prime.is_empty() {
                return Vec::new();
            }

            if stopped_by_x_parent {
                t_prime = zero_from_row(&t_prime, l + 2);
            }

            x_prime.push(t_prime);
            cursor += 1;

            last_l = l;
            last_stopped_by_x_parent = stopped_by_x_parent;

            if stopped_by_x_parent {
                x_end = cursor;
                break;
            }
        }

        // Determine if we keep y and the original X
        let mut keep_case1 = false;
        if x_end < cols.len() {
            let frc = first_row_column(z[0], n);
            keep_case1 = col_geq(&cols[x_end], &frc);
        }

        let mut keep_case2 = false;
        if last_l >= 0 && x_end > 0 && x_end - 1 < cols.len() {
            if cols[x_end - 1][last_l as usize] == 0 && ancestor.parent_is_column(x_end - 1, last_l as usize, y_idx) {
                keep_case2 = true;
            }
        }

        let keep_case3 = last_stopped_by_x_parent
            && (last_l + 1) < n as i32
            && x_end > 0
            && (x_end - 1) < cols.len()
            && cols[x_end - 1][(last_l + 1) as usize] > 0;

        let keep_original_yx = keep_case1 || keep_case2 || keep_case3;

        if keep_original_yx {
            // Keep y and x_prime after x, original X remains
            let mut new_cols: Columns = Vec::with_capacity(cols.len() + x_prime.len());
            new_cols.extend_from_slice(&cols[..iu + 1]);
            new_cols.extend(x_prime.iter().cloned());
            new_cols.extend_from_slice(&cols[iu + 1..]);
            cols = new_cols;
        } else {
            // Replace y..X with x_prime only
            let mut new_cols: Columns = Vec::with_capacity(cols.len());
            new_cols.extend_from_slice(&cols[..iu + 1]);
            new_cols.extend(x_prime.iter().cloned());
            new_cols.extend_from_slice(&cols[x_end..]);
            cols = new_cols;
        }

        idx -= 1;
    }

    cols
}

// ============================================================
// Standard BMS → Triangular BMS
// ============================================================

pub fn bms_to_triangular(m: &Matrix) -> Matrix {
    let mut cols = m.clone();
    if cols.is_empty() {
        return Vec::new();
    }
    let n = cols[0].len();
    if n < 2 {
        return Vec::new();
    }

    let mut idx: usize = 0;
    let mut steps: i64 = 0;
    const STEP_LIMIT: i64 = 100000;

    while idx < cols.len() {
        steps += 1;
        if steps > STEP_LIMIT {
            return Vec::new();
        }

        let x = cols[idx].clone();
        let k = last_positive_row(&x);
        if k >= n as i32 - 1 {
            idx += 1;
            continue;
        }

        let y = increment_prefix(&x, k + 1);
        let z = increment_row(&y, k + 2);

        let x_start = idx + 1;
        if x_start >= cols.len() || col_less(&cols[x_start], &z) {
            idx += 1;
            continue;
        }

        let mut x_end = x_start;
        while x_end < cols.len() && col_geq(&cols[x_end], &z) {
            x_end += 1;
        }

        let ancestor = AncestorIndex::new(&cols);
        let mut x_prime: Columns = Vec::new();

        for cursor in x_start..x_end {
            let t = cols[cursor].clone();

            let mut l: i32 = -1;
            for row in 0..=(k as usize + 1) {
                if ancestor.has_ancestor_column(cursor, row, idx) {
                    l = row as i32;
                }
            }
            if l < 0 {
                return Vec::new();
            }

            let is_last = cursor == x_end - 1;
            if is_last {
                if l < 0 || (l as usize) >= n {
                    return Vec::new();
                }
                if ancestor.parent_is_column(cursor, l as usize, idx) && t[l as usize] == 0 {
                    l -= 1;
                }
            }
            if l < 0 {
                return Vec::new();
            }

            let t_prime = increment_prefix(&t, l);
            x_prime.push(t_prime);
        }

        // comparison matrix: (y, ...x_prime, (y[0]+1,0,...,0))
        let mut insertion: Columns = Vec::new();
        insertion.push(y.clone());
        insertion.extend(x_prime.iter().cloned());

        let mut comparison = insertion.clone();
        comparison.push(first_row_column(y[0] + 1, n));

        let remainder: Columns = cols[x_end..].to_vec();

        // Always delete X
        cols.drain(x_start..x_end);

        // Insert if comparison > remainder (lexicographic tuple-of-tuples)
        if seq_cmp(&comparison, &remainder) > 0 {
            let mut new_cols: Columns = Vec::with_capacity(cols.len() + insertion.len());
            new_cols.extend_from_slice(&cols[..x_start]);
            new_cols.extend(insertion.iter().cloned());
            new_cols.extend_from_slice(&cols[x_start..]);
            cols = new_cols;
        }

        idx += 1;
    }

    cols
}
